cpp/src/parquet/arrow/arrow-reader-writer-test.cc

```diff
@@ -2420,6 +2420,81 @@ TEST(TestArrowWriterAdHoc, SchemaMismatch) {
   ASSERT_RAISES(Invalid, writer->WriteTable(*tbl, 1));
 }
 
+// ----------------------------------------------------------------------
+// Tests for directly reading DictionaryArray
+class TestArrowReadDictionary : public ::testing::TestWithParam<const char*> {
+ public:
+  void SetUp() override {
+    InitFromJSON(GetParam());
+    ASSERT_NO_FATAL_FAILURE(
+        WriteTableToBuffer(expected_dense_, expected_dense_->num_rows() / 2,
+                           default_arrow_writer_properties(), &buffer_));
+
+    properties_ = default_arrow_reader_properties();
+  }
+
+  void InitFromJSON(const char* jsonStr) {
+    auto dense_array = ::arrow::ArrayFromJSON(::arrow::utf8(), jsonStr);
+    expected_dense_ = MakeSimpleTable(dense_array, /*nullable=*/true);
+
+    ::arrow::StringDictionaryBuilder builder(default_memory_pool());
+    const auto& string_array = static_cast<const ::arrow::StringArray&>(*dense_array);
+
+    for (int64_t i = 0; i < string_array.length(); ++i) {
+      if (string_array.IsNull(i)) {
+        ASSERT_OK(builder.AppendNull());
+      } else {
+        ASSERT_OK(builder.Append(string_array.GetView(i)));
+      }
+    }
```

Comment:
can you use AppendArray instead? 

https://github.com/apache/arrow/blob/master/cpp/src/arrow/array/builder_dict.cc#L243